movement/watch_faces/complication/calculator_face.c

```diff
@@ -0,0 +1,345 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+#include "calculator_face.h"
+
+static void show_fn(calculator_state_t *state, uint8_t subsecond);
+
+void calculator_face_setup(movement_settings_t *settings, uint8_t watch_face_index, void ** context_ptr) {
+    (void) settings;
+    (void) watch_face_index;
+
+    if (*context_ptr == NULL) {
+        *context_ptr = malloc(sizeof(calculator_state_t));
+        memset(*context_ptr, 0, sizeof(calculator_state_t));
+        // Do any one-time tasks in here; the inside of this conditional happens only at boot.
+    }
+    // Do any pin or peripheral setup here; this will be called whenever the watch wakes from deep sleep.
+}
+
+static void calculator_face_display_number(double num) {
+    char buf[9] = {0};
+    bool negative = num < 0;
+    int max_digits = negative ? 5 : 6;
+
+    printf("%f\n", num);
+    if (negative) {
+        num = -num;
+    }
+
+    // Can we reasonably represent this number without a decimal point?
+    if (num == 0 || (num >= 0.5 && fabs(num - (int)num) < 0.0001)) {
+        if (floor(log10(num)) + 1 <= max_digits) {
+            if (negative) {
+                sprintf(buf, "  -%-5d", (int)round(num));
+            } else {
+                sprintf(buf, "  %-6d", (int)round(num));
+            }
+            watch_clear_colon();
+            watch_display_string(buf, 2);
+            return;
+        }
+    }
+
+    // Is this a floating point number where scientific
+    // notation won't get us much? (i.e. between 0.1 and 1)
+    if (num < 1 && num >= 0.0999) {
+        // Display as boring floating point number... (e.g. 0.25)
+        int digits = (int)round(num * 10000);
+        sprintf(buf, "   0%04d", digits);
+        if (negative) {
+            buf[2 ] = '-';
+        }
+        watch_set_colon();
+        watch_display_string(buf, 2);
+        return;
+    }
+
+    // Fall back to scientific notation
+
+    // Calculate exponent
+    int exponent = 0;
+    while (num < 1) {
+        num *= 10;
+        --exponent;
+    }
+
+    while (num >= 10) {
+        num /= 10;
+        ++exponent;
+    }
+
+    if (exponent < -9) {
+        sprintf(buf, "  small ");
+        watch_clear_colon();
+        watch_display_string(buf, 2);
+        return;
+    }
+
+    if (exponent > 39) {
+        sprintf(buf, "   big  ");
+        watch_clear_colon();
+        watch_display_string(buf, 2);
+        return;
+    }
+
+    sprintf(buf, "%2d%c%05d", exponent, negative ? '-' : ' ', (int)round(num * 10000));
+    watch_set_colon();
+    watch_display_string(buf, 2);
+}
+
+#define C (s->stack[s->stack_index])
```

Comment:
This was a bit of an experiment at writing code the way I usually wouldn't :)

(i.e. using macros and making everything deal with the global state instead of having clean functions)

Because the scope is so small, I _think_ it reads better.